### PR TITLE
videotoolbox: add --hwdec=videotoolbox-copy for h/w accelerated decoding with video filters

### PR DESCRIPTION
**Pull request by tmm1**
```
$ mpv --msg-level=all=v --hwdec=videotoolbox-copy -vf lavfi=vflip file.mp4
...
[vd] Opening video decoder lavc:h264
[vd] Probing 'videotoolbox-copy'...
[vd] Trying hardware decoding.
[vd] Selected video codec: H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 [lavc:h264]
Opening video filter: [lavfi graph=vflip]
[lavfi] Setting option 'graph' = 'vflip' (flags = 0)
...
Using hardware decoding (videotoolbox-copy).
[vd] Decoder format: 1280x720 nv12 bt.709/limited CL=mpeg2/4/h264
[lavfi] lavfi: create graph: 'vflip'
[ffmpeg] src: w:1280 h:720 pixfmt:nv12 tb:1/1000000 fr:0/1 sar:1/1 sws_param:
[vf] Video filter chain:
[vf]   [in] 1280x720 nv12 bt.709/limited CL=mpeg2/4/h264
[vf]   [lavfi] "lavfi.00" 1280x720 nv12 bt.709/limited CL=mpeg2/4/h264
[vf]   [out] 1280x720 nv12 bt.709/limited CL=mpeg2/4/h264
VO: [opengl] 1280x720 nv12
[cplayer] VO: Description: Extended OpenGL Renderer
...
```

### CODE REVIEW

File: video/decode/videotoolbox.c [hunk 1]
```diff
@@ -24,8 +24,28 @@
...
+    case AV_CODEC_ID_H263:
+    case AV_CODEC_ID_MPEG1VIDEO:
+    case AV_CODEC_ID_MPEG2VIDEO:
+    case AV_CODEC_ID_MPEG4:
+        break;
+    default:
+        return HWDEC_ERR_NO_CODEC;
+    }
+    return 0;
+}
```
Comment (wm4) [inline video/decode/videotoolbox.c]: Should be shared with the direct probe function. (Or rather, you could make the direct probe function call this function.)

File: video/decode/videotoolbox.c [hunk 2]
```diff
@@ -100,10 +123,93 @@ static int init_decoder(struct lavc_ctx *ctx, int w, int h)
...
+
+    int err = av_videotoolbox_default_init(ctx->avctx);
+    if (err < 0) {
+        print_videotoolbox_error(ctx->log, MSGL_ERR, "failed to init videotoolbox decoder", err);
+        return -1;
+    }
+
+    return 0;
+}
+
```
Comment (wm4) [inline video/decode/videotoolbox.c]: Can you try to also reduce code duplication here? (I realize it's not much code, and the pixel format selection seems like a bit of a pain.)

**Comment by wm4** [comment]
Made some comments, also also manpage changes are missing. Otherwise looks good to me.

**Comment by tmm1** [comment]
> Should be shared with the direct probe function. (Or rather, you could make the direct probe function call this function.)

Done.

> Can you try to also reduce code duplication here?

Done.

> manpage changes are missing

Fixed.

**Comment by wm4** [comment]
Thanks, merged. Eventually I'd like to get rid of the sort-of duplicated copy code too, but for now this will do.

**Pull request closed without merging**